### PR TITLE
Fix `_legalize_path` types

### PR DESCRIPTION
**Pull request by snejus**
## Description

Part 2 of the work fixing types in `beets.util.__init__` #5215.

Mypy was not happy here because `_legalize_stage` function implementation concatenates `path` and `extension` parameters, implying that their types need to match.

You can see that initially `path` parameter was defined as a `str` while `extension` was `bytes`.

In reality, depending on the `fragment` parameter value, `extension` was sometimes provided as a `str` and sometimes as `bytes`. The same parameter decided whether `path` gets converted into `bytes` within `_legalize_stage` implementation. No surprise that mypy was confused here.

`_legalize_stage` is only used within `Item.destination` method implementation which is where `fragment` is defined. I determined that the `fragment` parameter controls the form of the output path:

- `fragment=False` returned absolute path as **bytes** (default)
- `fragment=True` returned path relative to the library directory as **str**.

Given the above, the change

1. Renames `fragment` parameter to `relative_to_libdir` for clarity
2. Makes `Item.destination` to return the same type in both cases. I picked `bytes` since that's the type that majority of the code using this method expects.

   I converted the output path to `str` for the code that has been expecting a string there.

3. Decouples `_legalize_stage` and `_legalize_path` implementations from the `relative_to_libdir`. The logic now uses `str` type only.

### CODE REVIEW

**Comment by snejus** [comment]
@wisp3rwind this is ready for a review - trying to revive this work

**Review by wisp3rwind — COMMENTED** [review]
Nice seeing some progress here!

I think it's a great observation that there's only really one caller to `legalize_path` (i.e. `Item.destination`), which makes it straightforward to adapt the signature of `legalizat_path` to get rid of all its weirdness. This seems like the right thing to do!

It seems correct that the input to `legalize_path` should always be Unicode: The source it always a template evaluation, which is performed using Unicode strings, rather than something filesystem-specific or bytes. The output is most naturally `bytes`, since it's our canonical representation for things supposed to be passed to the filesystem.

At a relatively superficial glance, this PR looks reasonable to me (the one inline comment that I left is probably out-of-scope).

One thing I'm unsure about is the use `truncate` (but I'm not sure it made sense before, either): If the purpose of truncation is to deal with filesystem limitations, this should model as closely as possible what matters to the underlying filesystem. Previously, truncation could be performed on `str` or `bytes` depending on the `fragment` flag (notably, not depending on platform or filesystem). Now, it's always based on Unicode code points: In one sense, this is an improvement, since the result will always be valid UTF-8. On the other hand, the byte-length of the result is not bounded by the specified length. Ideally, I feel like this should attempt to truncate at Unicode graphemes boundaries, but measure length in terms of encoded bytes or Unicode code points depending on platform/filesystem.

It's also unexpected that the previous implementation used `displayable_path` in the legalization procedure: That is meant to remove non-printable characters, but does that align with what makes sense to be part of a filename? Maybe it would be interesting to check the `git blame` whether it has some rationale for its use here?

In any case, the previous solution doesn't seem to be ideal either (or, at least, it lacks sufficient documentation to make sense of it for me), so perfecting the truncation might be worth opening a tracking issue for, but out-of-scope in this PR. Do we have any bug reports that might be related to path truncation and legalization?

File: beets/library.py [hunk 1]
```diff
@@ -1064,26 +1064,20 @@ def move(
...
 
-        fragment makes this method return just the path fragment underneath
-        the root library directory; the path is also returned as Unicode
+        the path is also returned as Unicode
         instead of encoded as a bytestring. basedir can override the library's
         base directory for the destination.
         """
         self._check_db()
-        platform = platform or sys.platform
         basedir = basedir or self._db.directory
```
[the remark this answers is not in the document]
Comment (snejus) [reply beets/library.py:1078]: If only it was so simple... 😅 

image

`_check_db` return type says `Database`, while in reality this returns an instance of `Library`, which indeed has `directory` and `path_formats` attributes.

This entire design with an optional `_db` attribute has generally been driving me crazy 🤯.  It's a bit of a hard nut to crack - it's hardly possible to track down how and where `_db` is used because it's just about everywhere in the models code 😁 

I'll see whether I can adjust the type `_check_db` returns to `Library` - I remember when I tried doing it last time I gave up 🤣

[the remark this answers is not in the document]
Comment (wisp3rwind) [reply beets/library.py:1078]: Something like (untested, I might have mixed up the syntax):

```python
D = TypeVar("D")

class Model(Generic[D]):
    def __init__(self, db: D):
        self._db = db
        ...

    def _check_db(self) -> D:
        ...

class LibModel(Model[D]):
    ...

class Item(Libmodel[Library]):
    ...
```

In Rust, these would "associated types" on a trait. In beets, we do something similar in `dbcore.types`.

**Comment by snejus** [comment]
> It's also unexpected that the previous implementation used `displayable_path` in the legalization procedure: That is meant to remove non-printable characters, but does that align with what makes sense to be part of a filename? Maybe it would be interesting to check the `git blame` whether it has some rationale for its use here?

I followed your suggestion and had a look at the history of this functionality: [this PR](https://github.com/beetbox/beets/pull/1361) and [this issue](https://github.com/beetbox/beets/issues/1418) may be the most informative. Adrian nicely summarised the point of having to use bytes in [this comment](https://github.com/beetbox/beets/issues/1418#issuecomment-92215735):

> Thank you! This is related to https://github.com/beetbox/beets/issues/496, where truncation was also creating invalid filenames by bypassing the replace rules.

> To resolve this, you could imagine doing one of two things:

> Do the truncation on pre-encoded, Unicode character strings.
Somehow ensure, after truncation on the byte string, that we trim partial encoded characters.
The trouble with option 1 is that, on Unix, paths are byte strings—so a filesystem's maximum length is specified in bytes, not characters. (Not sure about Windows; it's probably characters there.) And the difficulty with 2 is that finding the character boundaries amounts to re-decoding back to Unicode.

See the rest of the thread around this comment. It seems like this may be less relevant in Python 3 ;)

**Comment by snejus** [comment]
All of this functionality was written back in the day when some useful libraries did not exist yet. Take [pathvalidate](https://pathvalidate.readthedocs.io/en/latest/pages/introduction/index.html), for example, which could be helpful for us here.

File: beets/library.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (snejus) [reply beets/library.py:1078]: I see! How would it help to make sure that the `_db` attribute is non-optional?

[the remark this answers is not in the document]
Comment (snejus) [reply beets/library.py:1078]: Ignore my question - I initially failed to understand your example properly. Now I see what's going on 😅

[the remark this answers is not in the document]
Comment (snejus) [reply beets/library.py:1078]: Though, we know that the `_db` attribute always has the `Library | None` type. Is there any need to have a type variable there in such case?

[the remark this answers is not in the document]
Comment (snejus) [reply beets/library.py:1078]: It worked wonderfully: see this PR: https://github.com/beetbox/beets/pull/5545

**Comment by wisp3rwind** [comment]
I've had another look, sorry for letting this slide for so long!

I don't think the truncation issues just disappear on Python 3: Consider [some filesystem with a 255-byte filename length limit](https://en.wikipedia.org/wiki/Comparison_of_file_systems#Limits) and UTF8 filename encoding. Now:

- The old code would encode to bytes before truncating. Thus, the filename would certainly obey the 255-bytes limit, but it might result in invalid Unicode if splitting code points or graphemes.
- The new code would truncate the Unicode strings. Thus, the result is certainly valid UTF8, but it might still split graphemes in unexpected ways (but that issue existed before, so I don't think it needs to be fixed in this PR). However, the result might exceed the 255 bytes limit, since each code point can encode to 1-4 bytes, giving a 1020 bytes filename in the worst case.

That is, for some combination of long filenames, non-latin characters in the filename and specific filesystems, this PR seems to be a regression.

Maybe, you could add a test case that obtains the maximum filename length for the filesystem used on the CI runner, generates a filename with exactly this many 2-byte UTF8 characters, and verifies that the resulting sanitized filename is sufficiently truncated?

For reference, [this](https://github.com/beetbox/beets/commit/953291f7366e00a0e473c458e56f20794b86a284) is where pathname truncation was introduced in the first place.

> All of this functionality was written back in the day when some useful libraries did not exist yet. Take [pathvalidate](https://pathvalidate.readthedocs.io/en/latest/pages/introduction/index.html), for example, which could be helpful for us here.

Looks like `pathvalidate` might indeed solve this problem, see the `max_len` and `fs_encoding` parameters of `pathvalidate.sanitize_filename` (it doesn't handle graphemes, though: https://github.com/thombashi/pathvalidate/blob/68b004fa15cf696285a6ae591ef318aca69c64fa/pathvalidate/_common.py#L158).

We could either
- use `pathvalidate`, or
- if we want to avoid any unnecessary changes to the validation logic and adding another dependency, just use the same truncation "trick" that is used by `pathvalidate` with the `ignore` option when re-decoding?

**Comment by snejus** [comment]
I see what you mean. Could we not simply encode the filename before truncation, and then decode it back?

**Comment by snejus** [comment]
@wisp3rwind feel free to have another look. Truncation is now done in bytes, and I added tests for the following scenario:

> Care should be taken not to incorrectly split UTF8 chars

I also simplified the entire legalization workflow and would like to hear your input regarding two things:

1. See 5b43bf0: our `MAX_FILENAME_LENGTH` of 200 is lower than the minimum any OS may report (255 or 256), so that check is redundant, as 200 will be always be less. Do you reckon we should instead trust the value that the system reports and use it instead of comparing it to our limit of 200?
2. See 9aa38a5 and the last test case. I found this behaviour to be counter-intuitive and thought that we should always respect user replacements and just truncate the path again if that's needed. This also made the logic overly complex. What do you think?

**Review by wisp3rwind — CHANGES_REQUESTED** [review]
> @wisp3rwind feel free to have another look. Truncation is now done in bytes, and I added tests for the following scenario:
> 
> > Care should be taken not to incorrectly split UTF8 chars

Had a very quick look at the truncation step, see the inline comment.

> I also simplified the entire legalization workflow and would like to hear your input regarding two things:
> 
>     1. See [5b43bf0](https://github.com/beetbox/beets/commit/5b43bf0c5085deda0592dd47513bcd58881f96ba): our `MAX_FILENAME_LENGTH` of 200 is lower than the minimum any OS may report (255 or 256), so that check is redundant, as 200 will be always be less. Do you reckon we should instead trust the value that the system reports and use it instead of comparing it to our limit of 200?

Not sure I have a clear opinion here: A quick glance at Wikipedia and stackoverflow suggests that it's increasingly unlikely to encounter smaller file name limits on any modern-ish system. On the other hand, the current version of `get_max_filename_length` should be tested by time. The complexity in `get_max_filename_length` also doesn't cause additional complexity elsewhere.

>     2. See [9aa38a5](https://github.com/beetbox/beets/commit/9aa38a5b048e9a4a1747f27b7cd278dd11104f17) and the last test case. I found this behaviour to be counter-intuitive and thought that we should always respect user replacements and just truncate the path again if that's needed. This also made the logic overly complex. What do you think?

Looking (very superficially) at https://github.com/beetbox/beets/pull/1361 and https://github.com/beetbox/beets/issues/496, it seems that (part of?) the reasoning behind the old code is to prevent the generation of illegal/hard-to-work-with paths after the final truncation  such as trailing spaces or dots. (Probably most relevant for Windows, which forbids more characters in paths.) Thus, it's not so much an issue about user vs. default replacements, but rather that the final step needs to be a replacement step, not a truncation step. Falling back to using the default replacements improves the user experience in cases where the the user replacements are messed up: Instead of an abort with `OSError`, the action will complete successfully (not messing up the database) and yield and informative error message. Not sure how often people experience that, but I could imagine that this actually happens occasionally when configuring new replacements.

This should probably be documented in more detail in the code comments, up to now I also never fully understood why the code is written as-is.

File: beets/util/__init__.py [hunk 2]
```diff
@@ -723,105 +725,51 @@ def sanitize_path(path: str, replacements: Replacements | None = None) -> str:
...
-    made the string longer again after it was truncated); the
-    application should probably log some sort of warning.
+    If we end up truncating a unicode character in the middle (rendering it invalid),
+    it is removed:
+
+    >>> s = "🎹🎶"  # 8 bytes
+    >>> truncate_str(s, 6)
+    '🎹'
     """
+    return os.fsencode(s)[:length].decode(errors="ignore")
```
[the remark this answers is not in the document]
Comment (snejus) [reply beets/util/__init__.py:738]: Nice catch!

**Comment by wisp3rwind** [comment]
> > Looking (very superficially) at #1361 and #496, it seems that (part of?) the reasoning behind the old code is to prevent the generation of illegal/hard-to-work-with paths after the final truncation such as trailing spaces or dots. (Probably most relevant for Windows, which forbids more characters in paths.) Thus, it's not so much an issue about user vs. default replacements, but rather that the final step needs to be a replacement step, not a truncation step. Falling back to using the default replacements improves the user experience in cases where the the user replacements are messed up: Instead of an abort with `OSError`, the action will complete successfully (not messing up the database) and yield and informative error message. Not sure how often people experience that, but I could imagine that this actually happens occasionally when configuring new replacements.
> 
> Okay, now I see what this was for!! So maybe, as the last step, we should sanitize the path with default replacements in every case, regardless of the length? I imagine we never want to have any invalid characters there.

Some reasons are in https://github.com/beetbox/beets/issues/496#issuecomment-75701952 and the preceding comment.

**Comment by snejus** [comment]
Have read the comments and reverted it back as I don't see a simple solution here. 

I thought we could selectively apply the relevant replacements for Windows and Unix systems. However, I see in code comments that Windows replacements should always be applied due to samba shares.

**Comment by snejus** [comment]
> On the other hand, the current version of get_max_filename_length should be tested by time.

Could you clarify what did you mean here?

**Comment by snejus** [comment]
@wisp3rwind I've undone my length adjustment now, let me know if there's anything missing.

I'm keen to finally get this out of the way!

**Comment by wisp3rwind** [comment]
There's a bunch of broken encoding-related tests on CI now, can you have a look?

Also, you previously removed a few tests from `test_library.py`. I think with the latest iteration of this PR; they can be added back because the behavior should be the same as it was originally.

**Comment by snejus** [comment]
> There's a bunch of broken encoding-related tests on CI now, can you have a look?

Didn't notice the failures - lookinga at them now.

> Also, you previously removed a few tests from test_library.py. I think with the latest iteration of this PR; they can be added back because the behavior should be the same as it was originally.

These have been replaced by `test.test_util.TestPathLegalization`!

**Comment by snejus** [comment]
@wisp3rwind tests are happy now :)

**Comment by wisp3rwind** [comment]
> > Also, you previously removed a few tests from test_library.py. I think with the latest iteration of this PR; they can be added back because the behavior should be the same as it was originally.
> 
> These have been replaced by `test.test_util.TestPathLegalization`!

Sorry, missed this due them also having moved to another file.

Would you mind rebasing this on top of master (with https://github.com/beetbox/beets/pull/5662)? Given that both touch path handling code, I think it'd be good to have a CI run to be sure that there are no unexpected interactions (even though I'd be surprised if there were).

**Comment by snejus** [comment]
Of course. I have earlier added branch protection to require branches to be up to date before merging anyways!

**Comment by wisp3rwind** [comment]
> Of course. I have earlier added branch protection to require branches to be up to date before merging anyways!

Yep, I've seen that, I think that's a great idea! Unfortunately, it appears that we can't use Github's merge queue for free, which would, as far as I understand it, help with not needing to do all those rebases manuall.
What I meant here is that I'd prefer to have it rebased before reviewing. Thanks!

File: test/test_library.py [hunk 3]
```diff
@@ -462,17 +452,6 @@ def test_destination_with_replacements(self):
...
-        self.lib.directory = b"base"
-        self.lib.replacements = [(re.compile(r"a"), "f")]
-        self.lib.path_formats = [("default", "$album/$title")]
-        self.i.title = "foo"
-        self.i.album = "bar"
-        replacements = [(re.compile(r"a"), "e")]
-        assert self.i.destination(replacements=replacements) == np(
-            "base/ber/foo"
-        )
-
```
Comment (wisp3rwind) [inline test/test_library.py:475]: Maybe, we want to keep this test? I read this more of a test about `destination()` honoring the `replacements` argument, rather than testing the replacement mechanism in detail.

File: test/test_util.py [hunk 4]
```diff
@@ -175,18 +170,86 @@ def test_bytesting_path_windows_removes_magic_prefix(self):
...
+            pytest.param(
+                [(re.compile(r"abcdX$"), "PRE_LONG")],
+                [(re.compile(r"PRE_$"), "POST_LONG")],
+                ("_PRE_", True),
+                id="too long repl both times force default ones to be applied",
+            ),
+        ],
+    )
+    def test_replacements(self, pre_trunc_repl, post_trunc_repl, expected):
+        replacements = pre_trunc_repl + post_trunc_repl
```
Comment (wisp3rwind) [inline test/test_util.py:253]:
```suggestion
        replacements = pre_trunc_repl + post_trunc_repl
        replacements = [(re.compile(pattern), repl) for pattern, repl in replacements]
```
and then remove all the explicit `re.compile`s above.

File: test/test_util.py [hunk 5]
```diff
@@ -175,18 +170,86 @@ def test_bytesting_path_windows_removes_magic_prefix(self):
...
+            ),
+            pytest.param(
+                [],
+                [(re.compile(r"abcdX$"), "POST_LONG")],
+                (":POST", False),
+                id="too long path after post-trunc repl is truncated",
+            ),
+            pytest.param(
+                [(re.compile(r"abcdX$"), "PRE")],
+                [(re.compile(r"PRE$"), "POST")],
```
Comment (wisp3rwind) [inline test/test_util.py:228]: I think it's great to formulate all of these tests in terms of pytest parameters :+1:

When reading the code, I do find it a little hard do wrap my head around what's going on in each case (for example, to verify that this includes the cases that were previously tested by the test in `test_library.py`). Maybe it would be helpful to write the steps out in a comment, and and combine both replacements in a single list (they're only used together, so the splitting here serves only as a kind of comment anyway). Something like
```py
# :abcdX --repl--> :PRE --trunc--> :PRE --repl --> :POST
pytest.param(
    [(r"abcdX$", "PRE"), (r"PRE$", "POST")],
    id=...,
)
```
I think that this might ease the mental gymnastics required to parse the code.

File: beets/util/__init__.py [hunk 6]
```diff
@@ -871,16 +855,21 @@ def command_output(cmd: list[BytesOrStr], shell: bool = False) -> CommandOutput:
...
     Windows), return `limit`.
     """
+    if length := beets.config["max_filename_length"].get(int):
+        return length
+
+    limit = MAX_FILENAME_LENGTH
     if hasattr(os, "statvfs"):
         try:
-            res = os.statvfs(path)
+            res = os.statvfs(beets.config["directory"].as_str())
```
Comment (wisp3rwind) [inline beets/util/__init__.py:872]: I'd prefer to stick to the previous way of passing `length` and `path` here through all callers. In that way, `beets.util` remains completely separated from other beets code (like library and configuration).

What was your reasoning behind this change? Just the simplification due passing around less arguments, or something more?

File: beets/util/__init__.py [hunk 7]
```diff
@@ -694,105 +696,87 @@ def sanitize_path(path: str, replacements: Replacements | None = None) -> str:
...
-        out[-1] = base + ext
 
-    return os.path.join(*out)
+def truncate_path(str_path: str) -> str:
+    """Truncate each path part to a legal length preserving the extension."""
+    max_length = get_max_filename_length()
+    path = Path(str_path)
+    parent_parts = [truncate_str(p, max_length) for p in path.parts[:-1]]
+    stem = truncate_str(path.stem, max_length - len(path.suffix))
+    return str(Path(*parent_parts, stem).with_suffix(path.suffix))
```
Comment (wisp3rwind) [inline beets/util/__init__.py:718]: This seems a bit messy due to all the conversions going on, but I suspect that this can't really be improved without further conversion to pathlib (which I think we shouldn't do in this PR to keep it as simple as possible).

Maybe, this could be improved in a follow-up PR which keeps the surface interface of `legalize_path`, using `str` paths, but internally immediately converts it to `Path` and passes that to all the subfunctions? I didn't test this, but I suspect that this might turn out to be a simplification.

**Review by wisp3rwind — CHANGES_REQUESTED** [review]
The legalization logic looks sound to me :+1:

I have a few doubts regarding how tests are organized, and regarding configuration handling, see the inline comments.

File: beets/util/__init__.py (hunk 6 above)
Comment (snejus) [reply beets/util/__init__.py:872]: The reasoning behind this change was primarily about proper encapsulation and responsibility allocation:

1. The max filename length is fundamentally an implementation detail of path legalization, not something higher-level functions should need to be aware of

2. The value is constant throughout the application's run, making it a perfect candidate for caching with `@cache`

3. The higher-level `Item.destination` function was just passing through a configuration value it doesn't directly use - it was acting as a messenger for several layers of function calls

4. Moving this calculation closer to where it's actually used (in `truncate_path`) follows the principle that knowledge should be localized where it's relevant

While I understand the concern about keeping util completely separated from beets code, I felt the improved encapsulation and removal of parameter passing through multiple layers justified the trade-off. The util module now handles its own dependencies rather than requiring callers to understand and provide implementation details.

File: test/test_library.py (hunk 3 above)
Comment (snejus) [reply test/test_library.py:475]: `destination()` does not have such parameter anymore, see 366a5fd

File: test/test_util.py (hunk 5 above)
Comment (wisp3rwind) [reply test/test_util.py]: Looking great!

**Comment by wisp3rwind** [comment]
For some reason, CI seems to be stuck. Not sure whether it really didn't start or just isn't reporting properly? I've seen something similar in another PR recently, where it recovered after another force-push.

File: beets/util/__init__.py (hunk 6 above)
Comment (wisp3rwind) [reply beets/util/__init__.py:872]: I would agree that this reasoning holds for the `max_filename_length` config value.

Regarding the argument to `statvfs`, I tend to disagree: Fixing this to the library directory means that `get_max_filename_length` really only applies to paths in the library. However, if the caller is `item.destination(basedir="/some/path/not/in/the/library")` the `statvfs` argument should be different. (The convert plugins seems to use this, for example. Arguably, the current implementation of `item.destination` is incorrect in this regard anyway: It should call `util.max_filename_length(basedir)`, not `maxlen = util.max_filename_length(db.directory)`.)

I'm not sure how many users would really be affected by these distinctions. In principle, any path could also cross several filesystem boundaries with different maximum filename lengths, and we don't handle that at all right now (that's probably fine, having additional mount points in the library directory seems like a very niche usecase).

File: test/test_library.py (hunk 3 above)
Comment (wisp3rwind) [reply test/test_library.py:475]: Sorry, must've missed that!

**Comment by snejus** [comment]
> For some reason, CI seems to be stuck. Not sure whether it really didn't start or just isn't reporting properly? I've seen something similar in another PR recently, where it recovered after another force-push.

Force-pushed again, let's see if it comes back

File: beets/util/__init__.py (hunk 6 above)
Comment (snejus) [reply beets/util/__init__.py:872]: I completely agree with you that we should be using the provided `basedir` for filename length check! And you hit the nail on the head with the following:

> the current implementation of item.destination is incorrect in this regard anyway

Before my changes in this PR I thought that the length was indeed calculated for the provided `basedir`; and only later, after I carefully stepped through the logic in detail, I realised that this isn't happening. At this point, I thought I had two options:

1. Make some assumptions about the intended functionality and 'fix' it. This would have had to apply to `art_destination` too, so the scope of the original intention to 'fix the types' would have quickly spiralled out of control 😅 
2. Restructure for clarity: keep the existing functionality but move this check away from the `destination` method. This now makes it more clear that the max filename length is just static, and that no arguments provided into `destination` have any effect here.

Presumably, this could have helped you to notice this issue as well?

**Comment by snejus** [comment]
Updated the PR description to document the changes. @wisp3rwind let me know if there's anything else needing addressing.

File: beets/util/__init__.py (hunk 6 above)
Comment (wisp3rwind) [reply beets/util/__init__.py:872]: To be honest, I'm not quite following your argument here. Wouldn't it also have been an option to keep the current function signatures as they are, and leave the decision on what to do about the argument of `max_filename_length()` for later?

It appears that there are a few possibilities regarding what the ideal final outcome should be:

- It turns out that we don't really need the whole `statvfs` code at all these days and can just hardcode a value. Then, a bit of code can just be removed.
- It turns out that we actually do need `statvfs`: Then, we should fix `destination` (and maybe also `art_destination`, but that seems much less likely to actually generate problematic filenames, since it only deals with subpaths).

The old code, as well as this PR, are somewhere in between.

I think (speaking for myself at least) that we don't currently understand this issue sufficiently well to decide which of these options is best. As you say, it's not really sensible to bloat this PR with figuring that out and refactoring a lot. However, my conclusion would then be to change the code as little as possible.

I have a hunch that it might actually matter to be accurate about filename length limits, in particular in situations where several wildly different file systems are involved. (For example, main library on some Linux FS, but a converted copy on a FAT/NTFS/network drive. At least, it's non-obvious to me why it shouldn't matter. Maybe we simply don't observe any issues with the current code because really long filenames are actually somewhat rare?)

---

In any case, I don't want to block this PR, overall, it improves the structure of the codebase, and I feel like we're not really getting anywhere with this discussion. Feel free to merge, but note that I disagree about the changes around filename length limits.

Comment (snejus) [reply beets/util/__init__.py:872]: I think the only thing that matters here is that __I did not change the behaviour__ - only restructured it to make it more clear what's going on.

And yes, I'd be more than happy to have it merged finally, as soon as you approve it!

**Comment by wisp3rwind** [comment]
Closes https://github.com/beetbox/beets/issues/1418
Superseded https://github.com/beetbox/beets/pull/1533

File: beets/util/__init__.py (hunk 6 above)
Comment (wisp3rwind) [reply beets/util/__init__.py:872]: Sorry, I should have explained better why I object: To me, the restructuring (of not passing around `basedir`) does not improve clarity — but it seems that this is ultimately personal opinion. Additionally, my impression is that the correct (although, not sure how relevant) solution is to use the `basedir` given in `destination`, and not the library directory. Thus, restructuring now would imply rolling back the restructuring if that change ends up being made.

What I'm going to do is to approve this PR, and open a new issue to track the remaining questions regarding filename length limits
